Add regtest documentation to book (#5714)
zcash/zcash: doc/book/src/dev/regtest.md

# Regtest

_Regtest_ ("regression test") is one of the three _network types_
supported by Zcash, the others being `testnet` and `mainnet`.
Regtest is an entirely local, self-contained mode -- your node or nodes
do not talk with peers out in the world. It gives you complete
control of the state of the blockchain and the sequence of events.
You start with an empty blockchain (just the genesis block, block 0).
Blocks can be mined instantly, and must be created explicitly. 

> You can run one or more regtest node on the same system.

Reviewer: node -> nodes. "...one ore more regtest nodes..."

The [RPC tests](https://github.com/zcash/zcash/tree/master/qa/rpc-tests)
use `regtest` mode (usually starting multiple nodes), but you may use it

> manually and interactively to learn, test, and  experiment. Most often

Reviewer 2: ```suggestion
manually and interactively to learn, test, and experiment. Most often
```

just one node is used in this case.

## Example session

> Here's a sample session (after you've built `bitcoind` and `bitcoin-cli`):

Reviewer 2: ```suggestion
Here's a sample session (after you've built `zcashd` and `zcash-cli`):
```

```
$ mkdir /tmp/regtest-datadir
$ cat <<EOF >/tmp/regtest-datadir/zcash.conf
regtest=1
rpcuser=u
rpcpassword=p
rpcport=18132
EOF
$ src/zcashd -daemon -datadir=/tmp/regtest-datadir
```

Watch `tmp/regtest-datadir/regtest/debug.log` to see what `zcashd` is doing.
It may also be useful to start `zcashd` with `-debug` to generate much
more logging. After several seconds, we can send RPCs to the node:

```
$ src/zcash-cli -datadir=/tmp/regtest-datadir getblockchaininfo
{
  "chain": "regtest",
  "blocks": 0,
  "initial_block_download_complete": false,
  "headers": 0,
  (...)
}
# Generate (mine) three blocks:
$ src/zcash-cli -datadir=/tmp/regtest-datadir generate 3
[
  "05040271f43f78e3870a88697eba201aa361ea5802c69eadaf920ff376787242",
  "0469f2df43dda69d521c482679b2db3c637b1721222511302584ac75e057c859",
  "0ab7a26e7b3b5dfca077728de90da0cfd1c49e1edbc130a52de4062b1aecac75"
]
$ src/zcash-cli -datadir=/tmp/regtest-datadir getblockchaininfo
{
  "chain": "regtest",
  "blocks": 3,
  "initial_block_download_complete": true,
  "headers": 3,
  (...)
}
$ src/zcash-cli -datadir=/tmp/regtest-datadir stop
Zcash server stopping
$ 
```
## Network upgrades

> Zcash has adoped a series of

Reviewer 2: ```suggestion
Zcash has adopted a series of
```

[network upgrades](https://github.com/zcash/zcash/blob/master/src/consensus/upgrades.cpp).
On `mainnet` and `testnet`, these activate at
fixed, known block heights ([example](https://github.com/zcash/zcash/blob/master/src/chainparams.cpp#L117)).
In `regtest` mode, you determine the activation heights. Upgrades may occur at
any height greater than 0, and multiple upgrades can occur at the same height. The upgrades
have a strict ordering (as shown in the upgrades source file); for example, Canopy can't
be activated before Blossom. 

You specify the upgrade heights using multiple `-nuparams=`_\<branch-id\>_ arguments.
(The branch IDs are available in the

> [upgrades file](https://github.com/zcash/zcash/blob/master/src/consensus/upgrades.cpp))

Reviewer 2: ```suggestion
[upgrades.cpp file](https://github.com/zcash/zcash/blob/master/src/consensus/upgrades.cpp).)
```

It's convenient to add these to the configuration file, for example:
```
$ cat <<EOF >>/tmp/regtest-datadir/zcash.conf
nuparams=76b809bb:1
nuparams=f5b9230b:5
EOF
```
(Alternatively, you can specify these on the `zcashd` command line.)

> You need not activate every upgrade explicitly; the example actives Sapling

Reviewer 2: ```suggestion
You need not activate every upgrade explicitly. The example activates Sapling
```

(branchID 76b809bb) at height 1; activating Sapling implies activating Overwinter, so this
is done automatically. Similarly, activating Heartwood at height 5
also simultaneously activates Blossom. Since no later upgrades are specified, none
of them will activate, regardless of height reached.

**IMPORTANT**: if you change the network upgrade heights from one
test run to the next, it's almost always necessary to start fresh
by removing the data directory, otherwise you'll encounter strange errors.

## Manual testing within an RPC (Python) test run

> Add the line `import time; time.sleep(999999)` (the units are seconds) somewhere

Reviewer 3: ```suggestion
It is often useful, either when debugging an issue or simply when you want to put
the node into a certain state, to use the RPC test framework to produce the desired
state and then be able to manually interrogate and modify that state using `zcash-cli` 
to execute RPC calls. An easy way to do that is as follows:

Add the line `import time; time.sleep(999999)` (the units are seconds) somewhere
```

within an RPC test to pause its execution at that point. Start the test, and then:

```
$ ps alx | grep zcashd
0  1000   96247   96246  20   0 1426304 123952 futex_ SLl+ pts/12   0:18 /g/zcash/src/zcashd -datadir=/tmp/test9d907s8a/96246/node0 -keypool=1 -discover=0 -rest -nuparams=5ba81b19:1 -nuparams=76b809bb:1 -debug=mempool -mempooltxcostlimit=8000
0  1000   96274   96246  20   0 744092 85568 -      RLl+ pts/12     0:05 /g/zcash/src/zcashd -datadir=/tmp/test9d907s8a/96246/node1 -keypool=1 -discover=0 -rest -nuparams=5ba81b19:1 -nuparams=76b809bb:1 -debug=mempool -mempooltxcostlimit=8000
$ 
```

> Now you can interact with the running test node by copy and pasting its

Reviewer 2: ```suggestion
Now you can interact with the running test node by copy-and-pasting its
```

> `-datadir` argument, for example:

Reviewer 3: If I recall correctly, you also often use `gdb` to connect to the zcashd process during the sleep, don't you? Could you add a note about how to do that somewhere in here?

Author: That's a very good idea, and you're right (I do that all the time). But debugging is a large enough topic that it could use its own document (which I'd very much like to write, and would be perfect for this book). If it's okay, let's restrict this document's audience to those who wish to perform "black box" (external) experiments using the RPC interface without delving into the code.

Reviewer 3: Okay. So the primary goal here of the sleep is to cause the RPC test to pause, so that you can interrogate the node's state using the RPC APIs, in the state that the RPC test has gotten the node to. That makes a lot of sense to me. I'll suggest some verbiage to explain the sleep.
